src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,143 @@
+#include "greedy_selector.hpp"
+
+#include <algorithm>
+#include <limits>
+#include <list>
+
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+// This subroutine scans the least desirable available choices until the
+// accumulated reject_cost() is lower than a required value.
+// If the thereby accumulated reject_desirability is not lower than a threshold
+// an iterator to the choice one past the subsequence to reject is returned.
+// Otherwise the end() iterator is returned, that the constraints are not satisfied
+// by any continuous subsequence starting at sorted_choices.begin().
+std::list<std::shared_ptr<TuningChoice>>::iterator sacrifice_choices(
+    std::list<std::shared_ptr<TuningChoice>>& sorted_choices, float required_cost_delta,
+    float acceptible_desirability_delta = -std::numeric_limits<float>::infinity()) {
+  float desirability_delta = 0.0f;
+  float cost_delta = 0.0f;
+
+  auto choice = sorted_choices.begin();
+  while (choice != sorted_choices.end()) {
+    if (cost_delta <= required_cost_delta) {
+      break;
+    }
+    desirability_delta += (*choice)->reject_desirability();
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.end();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
```

Comment:
abstract_selector.hpp line 18 made me think the other way around

Reply:
see above, mixing of choices and operations